src/Compilers/Core/Portable/SourceGeneration/Nodes/NodeStateTable.cs

```diff
@@ -239,29 +294,90 @@ public bool TryModifyEntries(ImmutableArray<T> outputs, IEqualityComparer<T> com
                 }
 
                 _states.Add(modified.ToImmutableAndFree());
+                RecordStepInfoForLastEntry(elapsedTime, stepInputs, overallInputState);
                 return true;
             }
 
-            public void AddEntry(T value, EntryState state)
+            public void AddEntry(T value, EntryState state, TimeSpan elapsedTime, ImmutableArray<(IncrementalGeneratorRunStep InputStep, int OutputIndex)> stepInputs, EntryState overallInputState)
             {
                 _states.Add(new TableEntry(value, state));
+                RecordStepInfoForLastEntry(elapsedTime, stepInputs, overallInputState);
             }
 
-            public void AddEntries(ImmutableArray<T> values, EntryState state)
+            public void AddEntries(ImmutableArray<T> values, EntryState state, TimeSpan elapsedTime, ImmutableArray<(IncrementalGeneratorRunStep InputStep, int OutputIndex)> stepInputs, EntryState overallInputState)
             {
                 _states.Add(new TableEntry(values, state));
+                RecordStepInfoForLastEntry(elapsedTime, stepInputs, overallInputState);
+            }
+
+            private void RecordStepInfoForLastEntry(TimeSpan elapsedTime, ImmutableArray<(IncrementalGeneratorRunStep InputStep, int OutputIndex)> stepInputs, EntryState overallInputState)
+            {
+                if (TrackIncrementalSteps)
+                {
+                    // We should have already recorded step information for all steps before the most recently recorded step.
+                    Debug.Assert(_steps.Count + 1 == _states.Count);
+
+                    TableEntry outputInfo = _states[^1];
+
+                    ArrayBuilder<(object Value, IncrementalStepRunReason OutputStatus)> stepOutputBuilder = ArrayBuilder<(object, IncrementalStepRunReason)>.GetInstance(outputInfo.Count);
```

Comment:
>ArrayBuilder<(object Value, IncrementalStepRunReason OutputStatus)> [](http://example.com/codeflow?start=20&length=67)

Consider using `var` since the type name is long and on the right hand side.